pkg/controller/user-cluster-controller-manager/ccm-csi-migrator/controller.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2021 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ccmcsimigrator
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"go.uber.org/zap"
+
+	"github.com/kubermatic/machine-controller/pkg/apis/cluster/common"
+	"github.com/kubermatic/machine-controller/pkg/apis/cluster/v1alpha1"
+	v1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	"k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1/helper"
+	"k8c.io/kubermatic/v2/pkg/version/kubermatic"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "ccm-csi-migrator"
+)
+
+type reconciler struct {
+	log          *zap.SugaredLogger
+	seedClient   ctrlruntimeclient.Client
+	userClient   ctrlruntimeclient.Client
+	seedRecorder record.EventRecorder
+	versions     kubermatic.Versions
+	clusterName  string
+}
+
+func Add(ctx context.Context, log *zap.SugaredLogger, seedMgr, userMgr manager.Manager, versions kubermatic.Versions, clusterName string) error {
+	log = log.Named(controllerName)
+
+	r := &reconciler{
+		log:          log,
+		seedClient:   seedMgr.GetClient(),
+		userClient:   userMgr.GetClient(),
+		seedRecorder: seedMgr.GetEventRecorderFor(controllerName),
+		versions:     versions,
+		clusterName:  clusterName,
+	}
+	c, err := controller.New(controllerName, userMgr, controller.Options{
+		Reconciler: r,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller %s: %v", controllerName, err)
+	}
+
+	// Watch for changes to Machines
+	if err = c.Watch(
+		&source.Kind{Type: &v1alpha1.Machine{}},
+		handler.EnqueueRequestsFromMapFunc(func(o ctrlruntimeclient.Object) []reconcile.Request {
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Name: clusterName,
```

Comment:
Does it make sense to read this value from the Machine object itself instead of passing it via flag? Machine objects have a label called `system/cluster` in `.spec.metadata.labels`.

I'm not sure when this label was added, but I think I've seen it for a while. On the other side, the current approach might be easier.

@moadqassem @mlavacca WDYT about this? I don't have strong opinions. I'm just thinking out loud.

Reply:
Yes, there is that label, and at the beginning, I used it to get the cluster name. But I moved into this approach because it is quite old and the key is hardcoded: https://github.com/kubermatic/kubermatic/blob/39d3d9e4a192b7af9d22f461feb82e44f3391d12/pkg/resources/machine/machinedeployment.go#L66
If someone in the future will remove or change it, there wouldn't be any reference to this controller. BTW, I guess it is a false problem, we can simply remove that hardcoding and refer to a common const. I think both approaches suit well, no strong opinion too.

Reply:
I don't have strong opinions about this, so I'm going to leave this up to you and @moadqassem. I'm also totally fine if we keep the current approach.